### ISSUE TITLE
[A11Y] Change color of the Views hidden fields for the Seven theme

### ISSUE THREAD

**Issue by philsward** [backdrop/backdrop-issues#5261]
**Description of the need**

Backdrop adds a [hidden] nomenclature to the end of a hidden field, however it still gets lost in the mix if there are a lot of fields, and the field names are long and wrap.

Example
![image](https://user-images.githubusercontent.com/3436110/135183668-d8204395-aa61-4d98-a320-94e1bed64e6f.png)
vs
![image](https://user-images.githubusercontent.com/3436110/135184665-60134793-6894-4ae5-8f3e-f6ac2e0393df.png)

**Proposed solution**

The proposed resolution is to change the color of the hidden items for better legibility between hidden and shown fields. I also recommend italicizing the hidden fields.

Example of real-world change on a Drupal site:

![image](https://user-images.githubusercontent.com/3436110/135183736-b064299c-3562-40b8-81b4-728815700124.png)

**Additional information**

The color used in my example is: rgb(114,131,145)

Recommended CSS addition (with slightly updated color recommendation):

**Files to be changed for CSS:**
* `core/modules/views_ui/css/views_ui.admin.theme.css`

**Files to be changed for moving `[hidden]` to the front of the line:**
* `core/modules/views_ui/views_ui.admin.inc`

```
.views-edit-view .views-field-excluded {
  color: rgb(160 160 160);
  font-style: italic;
}

```

**Draft of feature description for Press Release (1 paragraph at most)**

The Backdrop Seven theme now changes the color of hidden fields to better show which fields are hidden vs shown.

**Comment by bugfolder** [comment]
I like the idea of providing visual distinction (versus more words). 

Color contrast on the gray zebra striping seems low for this particular color choice. Also wonder about inconsistency of styling hyperlinks: light gray suggests something disabled; it might not be obvious that those elements could still be clicked/tapped to open.

Just spitballing more ideas: would an icon convey the exclusion as well? What about putting the icon at the beginning of the text, rather than tacking it on to the end?

**Comment by philsward** [comment]
@bugfolder the icon idea isn't a bad thought.

The color itself doesn't concern me as much as just being able to distinguish hidden vs shown.

I played with lighter colors and it became difficult for me to even read the hidden text. Darker or black seemed out of place to me but that's my personal taste.

**Comment by KinnariU** [comment]
I am working on this issue

**Comment by jenlampton** [comment]
I'm concerned about changing the color of the link here, because the blue color in the views interface has a specific meaning: "This is a link". In addition, the recommended gray color shown above is not accessable, so we should not be using that one. 

Because we already have an indicator that a field is hidden, I wonder if we shouldn't be finding a way to make that more useful, rather than adding more clutter in an attempt to do the same thing. I like the idea of adding an icon, but I also wonder if moving `[hidden]` to the beginning of the field name (rather than at the end) might solve the problem?

**Comment by philsward** [comment]
@jenlampton we can try moving hidden to the beginning but I fear it will also get lost.

I think having hidden at the front (will have to see it to judge whether I like it) and a color change will be intuitive enough for new folks to understand "This is hidden, oh and it's denoted as hidden with a color change". I highly doubt someone would exclude a field, only to see it grayed out and then honestly think they couldn't ever change it again.

The fix to this (just thought of it) is to change the hidden color back to the normal link color on hover which is an easy change. So, even if it is grayed out, hovering it will make it appear as a link again.

Mentioning accessibility, I've got a big picture idea for addressing accessibility as a whole, but haven't brought it up because I'm guessing it will go over about as well as a fart in an astronaut suit...

**Comment by jenlampton** [comment]
> we can try moving hidden to the beginning but I fear it will also get lost.

I'm still against changing the color, but what if we moved it to the beginning and also added an icon (we could do this with css)? Or moved it to the beginning and also added emphasis? (You mentioned the italics before, but I think adding emphasis is the opposite of what we want here...  but it would still be safer than a color change, imho)

**Comment by bugfolder** [comment]
I agree that we should be consistent with coloring hyperlinks.

Moving the hidden indicator to the beginning is, I think, an improvement over putting it at the end. I would find a boldly colored icon more eye-catching and less space-filling than the current "[hidden]" (which you have to read and parse before realizing that it's not part of the field name). But I don't have a good suggestion for what a good icon choice would be; perhaps further comments could chime in with ideas.

**Comment by philsward** [comment]
I would prefer just an icon to moving hidden to the front and also adding an icon. Field names can become extremely long especially with wrapping and adding [hidden] just makes it longer regardless where it sits.

If it's really that big of a personal preference, make it a setting for views that is disabled by default. Then use telemetry to find out "how many people turn this on".

It begs a much bigger rhetorical question of: "what needs to be done to get telemetry and a/b testing going so we can make informed decisions on UI and UX changes?"

**Comment by philsward** [comment]
@bugfolder If we go the icon route, then both hidden and viewable will need an icon.

How about something like this:

![image](https://user-images.githubusercontent.com/3436110/135683900-23dd4570-8819-4139-9a05-22819a370ddc.png)

**Comment by philsward** [comment]
Better example:

![image](https://user-images.githubusercontent.com/3436110/135684398-0b6aacc4-382d-4227-88b8-ced53de2154d.png)

**Comment by bugfolder** [comment]
I agree: clean and intuitive, and for the icon, the color difference is clear and appropriate.

If I had my druthers, I'd drop the "[hidden]", but I suspect that would make the meaning of the icon too mysterious for newcomers. But I really like the way it's possible to instantly discern what's visible and hidden, even when I'm not registering the individual words of the field.

(60-year-old eyes may be playing a role in this evaluation.)

**Comment by philsward** [comment]
Images pulled from: https://www.toptal.com/designers/colorfilter/

The gray contrasted with the blue, honestly isn't near as bad for colorblind as what I thought it would be...

![image](https://user-images.githubusercontent.com/3436110/135686093-acc77b97-c389-4dd5-bc3a-5f1d6891c203.png)

![image](https://user-images.githubusercontent.com/3436110/135686135-373a5ded-4751-49f0-9727-c17096aeb7af.png)

![image](https://user-images.githubusercontent.com/3436110/135686190-46d1a6ff-271f-46d3-b4d5-f0b8ed46ec7c.png)

![image](https://user-images.githubusercontent.com/3436110/135686236-294beaa7-c357-4d88-b8fe-9bbde24263e6.png)

**Comment by philsward** [comment]
Side Note: If the icon idea is a go, I think the best approach is to add the icon within its own span or div instead of using the ::before pseudo. Styling the ::before is causing weird issues in a sub-theme and placing it within its own element will give better styling control.

**Comment by philsward** [comment]
Potential blocker:

https://github.com/backdrop/backdrop-issues/issues/364

**Comment by olafgrabienski** [comment]
I'm quite skeptical if the use of (two) icons is a good idea. While the approach might have advantages in some use-cases (many fields, many of them hidden), the icons would clutter the interface without real need in other cases, i.e. on simple displays, and even when there are no hidden fields at all.

In my opinion, the `[hidden]` marker at the beginning would already be very helpful, and in a much less obtrusive way.

**Comment by philsward** [comment]
@olafgrabienski the reason for two icons is the same reason I'm against moving [hidden] to the beginning. An icon for each means that beginning column is standard all the way down. Otherwise we end up with:

Some text field
Some image field
[hidden] some path field
Some title field
Some body field
Some term field
[hidden] some random field
Some extra paragraph field

The point of the request is to make the fields area easier to read and understand, especially when there are a lot of fields. Moving [hidden] to the beginning will still make it difficult to read and understand if all of the field names line up at the beginning. Nothing is actually solved moving it from the end to the beginning.

I agree that 4-5 fields isn't a big deal. Am I the only one building views with 20+ fields and some are hidden for rewriting other fields? I use hidden fields a lot to deal with rewriting output, even on the most basic of views so it surprises me if others use it sparingly.

BTW, this isn't a new idea, nor can I take originality for the idea: https://www.drupal.org/project/views/issues/2757753

I would also like to point out the Drupal request to add an indicator for rewritten fields and I agree, this too would be a welcomed indicator.

**Comment by jenlampton** [comment]
> Nothing is actually solved moving it from the end to the beginning.

That's not true. The problem with having it at the end of the line is that the hidden line does not stand out from the one that's shown. It's especially bad when one field wraps to a second line, so the "Hidden" part that's usually at the end, is now on another line. Moving "Hidden" to the beginning of the line solves 2 problems. 

> I agree that 4-5 fields isn't a big deal. Am I the only one building views with 20+ fields and some are hidden for rewriting other fields?

No, I do this too :)

**Comment by jenlampton** [comment]
> I don't think hidden/visible fields is a common enough use-case to warrant the extra clutter that would be added to the UI 100% of the time.

here's a thought: What if the visible icon only appeared after the first hidden field was added to a view? I would be in support of both icons as long as the visible one didn't appear when all fields were visible. In fact, I'd also be in favor of removing "[hidden]" if both icons had a hover/title that said `Visible`/`Hidden`.

**Comment by philsward** [comment]
Interesting how some see information as clutter and others see a lack of information as clutter.

What about a views setting to enable/disable this as an option? I don't think there have been any comments on that idea yet. Good? Bad?

**Comment by bugfolder** [comment]
> I would also like to point out the Drupal request to add an indicator for rewritten fields and I agree, this too would be a welcomed indicator.

I like this too.

> [JL] I don't think hidden/visible fields is a common enough use-case to warrant the extra clutter that would be added to the UI 100% of the time.

> [PW] What about a views setting to enable/disable this as an option? 

I use hidden fields and field-rewriting a fair amount (the former to support the latter), so for me, they're a pretty common use case. So I would welcome a views setting to switch between the old way of doing things ("[hidden]" at the end) versus using distinct icons at the beginning of a line to indicate hidden or rewritten. If I had my druthers, I'd put no icon for fields that were neither hidden nor rewritten.

**Comment by jenlampton** [comment]
I also support an indicator for rewritten, but that's another issue ;)

I'll start on a PR that moves hidden to the front of the line, and wraps it
in a span we can use to both hide the text, and replace it with an icon
(for sighted users). I think a setting for "use icons for things" might be
unnecessary since most people seem to agree they are an improvement over
the text.

**Comment by olafgrabienski** [comment]
> I also support an indicator for rewritten, but that's another issue

Agreed! Having another indicator in mind, we can however already consider how it would fit the various suggestions, e.g. will we find a good 'rewritten' icon? Or, how does the two icons concept (hidden vs. visible) work when we also have a 'rewritten' indicator?

Moving hidden to the front and wrapping it in a span is an very good starting point, in my opinion. Looking forward to test the PR!

**Comment by philsward** [comment]
🥰 🎉

---

I'll get an issue going for the rewritten indicator and see if I can find some font awesome icons that might make sense as a starting point.

**Comment by klonos** [comment]
My initial reaction was to keep the link color as is (for the same reasons others have already mentioned), but move the `[hidden]` indicator as a plain-text suffix added to the link (instead of having it be part of the link as it is currently). A simple solution:

<img width=50% src="https://user-images.githubusercontent.com/2423362/136106286-779c6778-b7be-43d7-a663-a1ed7afa800a.png">

Having said that, I'm not against using icons...

> I like the idea of adding an icon, but I also wonder if moving `[hidden]` to the beginning of the field name (rather than at the end) might solve the problem?

> Moving the hidden indicator to the beginning is, I think, an improvement over putting it at the end.

I don't particularly like the idea of the `[hidden]` indicator as a prefix, as I find it to be messing with the scannability of the field names, and breaking consistency (the later being a personal opinion or matter of perception).

> If we go the icon route, then both hidden and viewable will need an icon.

Yes! 👍🏼 ...I find adding open/closed eye icons in front of _all_ fields to be more consistent. In other words, this:

> An icon for each means that beginning column is standard all the way down.

> If I had my druthers, I'd drop the "[hidden]", but I suspect that would make the meaning of the icon too mysterious for newcomers.

I'm ambivalent re that too 🤔 ...on the one hand, I'd like shorter field names; icon-only would indeed make things "mysterious" for people that are not familiar with this UI.

> The fix to this (just thought of it) is to change the hidden color back to the normal link color on hover which is an easy change. So, even if it is grayed out, hovering it will make it appear as a link again.

> I'd also be in favor of removing "[hidden]" if both icons had a hover/title that said `Visible`/`Hidden`.

We could also make this work on-focus, which would cover keyboard-based navigation, but on mobile devices or any tab-based device (w/o a cursor or pointing device) hover/focus events don't apply.

**Comment by klonos** [comment]
...I would much rather prefer it if we reserved the left side of the fields space reserved for drag handles, so that we don't have to open a separate dialog for reordering and removing fields. This is a mockup I put together [back in 2017](https://github.com/backdrop/backdrop-issues/issues/930#issuecomment-298113041):

![backdrop-issue930-mockup-delete-undo](https://cloud.githubusercontent.com/assets/2423362/20452238/4db02038-ae59-11e6-9e99-217ddacc439f.png)

**Comment by philsward** [comment]
> If we keep "[hidden]", and especially if we keep it at the end, I really like pulling it out of the link as you did; that makes it stand out. And if it's wrapped in a classed span, then fans of icons (such as myself) could make a contrib module with a little JS to do other magic rearranging.

>Come to think of it, could we add classes like hidden and rewritten and the like to the div that encloses the link that reflect what's been done to the field? That would give a lot of flexibility to customize with text, icons, blinking-gifs, whatever.

I would be OK with this approach so-long as like @bugfolder said, those of us who want to override it have better options to.

I'll be honest and say that when I opened this, I didn't realize how easy it was to sub-theme Seven, making some of these requests a bit moot if there is an easy way to override them.

**Comment by jenlampton** [comment]
I don't know how easy or hard it is to add classes to each of these rows, but I think it's a great idea and would give everyone room for their own solutions :) 

@klonos the only problem with the mockup that pulls `[hidden]` out of the link is that the text becomes black, giving those items more emphasis than the ones that are shown, instead of less. I almost (?) prefer the red strike-through text, do you remember what that was supposed to signify?

**Comment by olafgrabienski** [comment]
> My initial reaction was to keep the link color as is ..., but move the [hidden] indicator as a plain-text suffix added to the link

Excellent, so simple and so good!

> the only problem with the mockup that pulls [hidden] out of the link is that the text becomes black, giving those items more emphasis

We can make the [hidden] text a bit lighter than black (dark gray), so that the suffix doesn't stand out too much.

> I almost (?) prefer the red strike-through text, do you remember what that was supposed to signify?

Found it: Corresponding to https://github.com/backdrop/backdrop-issues/issues/930#issuecomment-261685910 it means "marked for deletion". Something along those lines was also my first impression. I don't think it's a good way to represent the "hidden" state.

**Comment by klonos** [comment]
> I almost (?) prefer the red strike-through text, do you remember what that was supposed to signify?

Yes, I do 🙂 ...the x icon/buttons to the right of each field were there in order to remove them, but that would not happen instantly; the field would instead be "marked for deletion" once the view would be saved. Notice how in the the mockup the x "remove" button has changed to a "revert"/"restore" button instead? ...that for cases where you may have accidentally clicked "delete" in one of the fields 😉

**Comment by olafgrabienski** [comment]
>> the mockup that pulls [hidden] out of the link is that the text becomes black, giving those items more emphasis

> We can make the [hidden] text a bit lighter than black (dark gray), so that the suffix doesn't stand out too much.

I've tested it in a mockup, and it turned out that a gray with a contrast ratio comparable to the link color looks very light, so I've also tried a gray with higher contrast (but still not black). See screenshots below for comparison.

---

**(a) Default black** (too dark?)

<img width="693" alt="backdrop-views-hidden-color-000000" src="https://user-images.githubusercontent.com/14188307/136251574-046b2dd6-e004-4c4e-b5b6-a1033da7839a.png">

---

**(b) Gray `#424242`** (my favorite)

<img width="693" alt="backdrop-views-hidden-color-424242" src="https://user-images.githubusercontent.com/14188307/136251581-ac9b3554-859c-41d6-b3ae-6a7c7196dbb2.png">

---

**(c) Gray `#707070`** (too light?)

<img width="693" alt="backdrop-views-hidden-color-707070" src="https://user-images.githubusercontent.com/14188307/136251584-23a9000a-91ab-43e2-b6d9-0f54854e72f4.png">

**Comment by philsward** [comment]
I think somewhere in the #4XXXXX range will work.

**Comment by RakeshPotnuru** [comment]
Hey everyone
Is this discussion still going on?

**Comment by stpaultim** [comment]
@RakeshPotnuru - We have lots of open discussions like this and we don't remember the status of each one. You can help us, but reading through the issue and summarizing:

1) What is the problem
2) What is the recommended solution or solutions
3) Suggesting what you think the next step should be. 

Often times, we just need someone to create a Pull Request based upon the discussion so far, with the understanding that once we have a Pull Request there will be additional feedback and we might need to make more changes. 

If you think you need a decision before you can write a Pull Request, then give us a summary of where the discussion is at and which decision needs to be made. We could bring this up at our next design/us meeting if we have a summary.

**Comment by RakeshPotnuru** [comment]
@stpaultim 
Yes, I understood the problem. But they are discussing which colour to use. So I asked.
